app-ng/src/utilities.lisp

```diff
@@ -4,36 +4,27 @@
 
 (in-package #:qvm-app-ng)
 
-(defun session-info ()
-  ;; Stub implementation for FORMAT-LOG, below. See app/src/utilities.lisp for the original.
-  "")
+(defun get-random-state (arg)
+  (etypecase arg
+    (null (qvm:seeded-random-state nil))
+    (unsigned-byte (qvm:seeded-random-state arg))))
+
+(defun hex-char-p (char &key (case-sensitive-p t))
+  (case (if case-sensitive-p char (char-downcase char))
```

Comment:
```
(find char "0123456789abcdef" :test (if sensitive #'char= #'char-equal))
```

Reply:
I originally implemented this with `member`, then switched to `case` for HIGH SPEED PERFORMANCE!  But yeah probably unnecessary and definitely "premature optimization"

Reply:
Maybe not even an optimization at all, come to think of it

Reply:
For the noobs: how is this more performant?

Reply:
Any reason not to use the standard `(digit-char-p char 16)`? Is case sensitivity needed?

Reply:
> For the noobs: how is this more performant?

Upon reflection, I'm not even sure it is. With "sufficiently smart" compilers like SBCL's Python, you never really know until you look at the compiled code / benchmark it.

But often, the compiler can generate faster code for `case` than it can for a call to `member` or `find` since all the cases are there for inspection at compile time. But maybe SBCL can generate as good a code for `member` / `find` when the sequence you are searching is static... something fun to look into!

Reply:
> Any reason not to use the standard `(digit-char-p char 16)`? Is case sensitivity needed?

I forgot about `digit-char-p`. I did consider `parse-integer` with radix 16, but decided it was too hacky :-).

As mentioned elsewhere, I do currently make use of the case-sensitivity, but being able to delete this function altogether and just using `digit-char-p` might be enough to tip the scales the other way.

Reply:
I was only calling `hex-char-p` from one place, so I just deleted it and switched to calling `find` there.

Reply:
> > For the noobs: how is this more performant?
> 
> Upon reflection, I'm not even sure it is. With "sufficiently smart" compilers like SBCL's Python, you never really know until you look at the compiled code / benchmark it.
> 
> But often, the compiler can generate faster code for `case` than it can for a call to `member` or `find` since all the cases are there for inspection at compile time. But maybe SBCL can generate as good a code for `member` / `find` when the sequence you are searching is static... something fun to look into!

I had some time on the flight back last week. What better way to pass the time than staring into the warm glow of an SBCL `DISASSEMBLE`? It turns out that SBCL is "sufficiently smart" to unroll and inline the membership tests when `MEMBER`'s list argument is static. What a time to be alive!

Here is the code SBCL generates for `(member c '(#\a #\b #\c))`, which is practically identical to the code it generates for `(case c ((#\a #\b #\c) t))` (aside from the return value).

```
CL-USER> (disassemble (lambda (c) (member c '(#\a #\b #\c))))
; disassembly for (LAMBDA (C))
; Size: 72 bytes. Origin: #x23DF8650                          ; (LAMBDA (C))
; 50:       498B5D10         MOV RBX, [R13+16]                ; thread.binding-stack-pointer
; 54:       48895DF8         MOV [RBP-8], RBX
; 58:       483D49610000     CMP RAX, 24905
; 5E:       750D             JNE L1
; 60:       488B15A9FFFFFF   MOV RDX, [RIP-87]                ; '(#\a #\b #\c)
; 67: L0:   488BE5           MOV RSP, RBP
; 6A:       F8               CLC
; 6B:       5D               POP RBP
; 6C:       C3               RET
; 6D: L1:   483D49620000     CMP RAX, 25161
; 73:       7418             JEQ L2
; 75:       483D49630000     CMP RAX, 25417
; 7B:       488B0D96FFFFFF   MOV RCX, [RIP-106]               ; '(#\c)
; 82:       BA17001020       MOV EDX, #x20100017              ; NIL
; 87:       480F44D1         CMOVEQ RDX, RCX
; 8B:       EBDA             JMP L0
; 8D: L2:   488B158CFFFFFF   MOV RDX, [RIP-116]               ; '(#\b #\c)
; 94:       EBD1             JMP L0
; 96:       CC0F             BREAK 15                         ; Invalid argument count trap
NIL
```

Three `CMP` instructions, nary a `CALL` nor a loop in site. Took me a wee minute to realize that the literal char values are spaced 2^8 apart and thus probably have a type tag in the lowest word, hence 24905 == `#\a`. According to the SBCL Internal's [docs](http://www.sbcl.org/sbcl-internals/Memory-Layout.html#Memory-Layout), this is must be the so-called [widetag](http://www.chiark.greenend.org.uk/doc/sbcl-doc/html/sbcl-internals/Type-tags.html#Widetags).

```
CL-USER> (- 25161 24905)
256
CL-USER> (ash 24905 -8)
97
CL-USER> (char-code #\a)
97
```

The magic happens in [`TRANSFORM-LIST-ITEM-SEEK`](https://github.com/sbcl/sbcl/blob/1fca8073122056053a974182c4ed60b3a1484535/src/compiler/seqtran.lisp#L433) in `seqtran.lisp`, and the optimization only kicks in for static lists with length less than `sb-c::*list-open-code-limit*` (128 by default).

Sadly `find` does not get the same treatment and results in a function call to `sb-kernel:%find-position`, which I guess is reasonable since `find` has work with arbitrary sequences.

Reply:
That's great indeed. This was the rationale for embedding the permutation look-up table into the code in https://github.com/rigetti/qvm/blob/b63eed331285446ea0f54453c8c62c8946e55c61/dqvm/src/permutation.lisp#L320

Reply:
😎